urlefa3b0a07b_jhngzhu_Database_RelationalOperator_tgz-pJ8-tests_ROTestJ8/src/relop/HashJoin.java

```diff
@@ -0,0 +1,309 @@
+package relop;
+
+
+import global.AttrOperator;
+import global.AttrType;
+import global.RID;
+import global.SearchKey;
+import heap.HeapFile;
+import index.HashIndex;
+import relop.FileScan;
+import relop.HashJoin;
+import relop.IndexScan;
+import relop.KeyScan;
+import relop.Predicate;
+import relop.Projection;
+import relop.Schema;
+import relop.Selection;
+import relop.SimpleSortMergeJoin;
+import relop.Tuple;
+
+/**
+ * Implements the hash-based join algorithm described in section 14.4.3 of the
+ * textbook (3rd edition; see pages 463 to 464). HashIndex is used to partition
+ * the tuples into buckets, and HashTableDup is used to store a partition in
+ * memory during the matching phase.
+ */
+public class HashJoin extends Iterator {
+  public final static int NUM_BUCKETS=5;
+  public static int hfid=0;
+  public Iterator left;
+  public Iterator right;
+  public boolean openning;
+  public boolean isHasNext;
+  public Tuple nextTuple;
+  public Tuple leftTuple;
+  public Tuple rightTuple;
+  public boolean isFirstCall;
+  public Predicate pred;
+  public HeapFile lefthf[];
+  public HeapFile righthf[];
+  public FileScan leftscan[];
+  public FileScan rightscan[];
+  public FileScan curright;
+  public HashTableDup hashTableDup;
+  public int rcol;
+  public int lcol;
+  public int ind;
+  public int leftind;
+  public int rightind;
+  public Tuple ltuples[];
+
+  /**
+   * Constructs a hash join, given the left and right iterators and which
+   * columns to match (relative to their individual schemas).
+   */
+
+  public HashJoin(Iterator left, Iterator right, Integer lcol, Integer rcol) {
+//    throw new UnsupportedOperationException("Not implemented");
+    lefthf=getHeapFile(left,lcol);
+    righthf=getHeapFile(right,rcol);
+    leftscan=getFileScan(lefthf,left.schema);
+    rightscan=getFileScan(righthf,right.schema);
+    this.left=left;
+    this.right=right;
+    this.lcol=lcol;
+    this.rcol=rcol;
+    this.schema=Schema.join(left.schema,right.schema);
+    pred=getPred(left.schema,right.schema,lcol,rcol);
+    openning=true;
+    isHasNext=false;
+    nextTuple=null;
+    leftTuple=null;
+    rightTuple=null;
+    isFirstCall=true;
+    prepareNext();
+  }
+
+  /**
+   * Gives a one-line explaination of the iterator, repeats the call on any
+   * child iterators, and increases the indent depth along the way.
+   */
+  public void explain(int depth) {
+//    throw new UnsupportedOperationException("Not implemented");
+    indent(depth);
+    System.out.println("This is a HashJoin Iterator");
+  }
+
+  /**
+   * Restarts the iterator, i.e. as if it were just constructed.
+   */
+  public void restart() {
+//    throw new UnsupportedOperationException("Not implemented");
+    //I dont know how to do this yet.
+    left.restart();
+    right.restart();
+    for(int i=0;i<NUM_BUCKETS;i++){
+      leftscan[i].close();
+      rightscan[i].close();
+      lefthf[i].deleteFile();
+      righthf[i].deleteFile();
+    }
```

Comment:
Recommendation generated by Amazon CodeGuru Reviewer. Leave feedback on this recommendation by replying to the comment or by reacting to the comment using emoji.

Similar code fragments were detected in the same file at the following lines: 96:101, 129:134.  
Refactoring can help improve code maintainability. Consider reducing duplicate code by extracting it into a separate method. You can then replace duplicated code with calls to this new method.